src/services/completions.ts

```diff
@@ -323,15 +323,25 @@ namespace ts.Completions {
             return result;
         }
 
+        /**
+         * Given a path ending at a directory, gets the completions for the path, and filters for those entries containing the basename.
+         */
         function getCompletionEntriesForDirectoryFragment(fragment: string, scriptPath: string, extensions: string[], includeExtensions: boolean, span: TextSpan, exclude?: string, result: CompletionEntry[] = []): CompletionEntry[] {
+            if (fragment === undefined) { fragment = "./"; }
+
+            fragment = normalizeSlashes(fragment);
+
+            const baseName = getBaseFileName(fragment);
+
+            // Remove the basename from our directory path
             fragment = getDirectoryPath(fragment);
-            if (!fragment) {
-                fragment = "./";
-            }
-            else {
-                fragment = ensureTrailingDirectorySeparator(fragment);
+
+            if (fragment === "") {
+                fragment = "." + directorySeparator;

```

Comment:
This function really should never get `undefined` as an argument for fragment, but I don't know of a way to assert that. So the check on line 330 is really a matter of null-checking. In particular, `normalizeSlashes()` on line 332 dereferences its argument without null-checking.

Regarding line 340, consider the case when `fragment` is `"foo"`. Then `getDirectoryPath(fragment) === ""` because it removes the basename part of the path. So we need to check again.

Really, `fragment` no longer refers to a path fragment after the assignment on line 337. Perhaps a new variable called `relativeOrAbsolutePath` should be initialized on the LHS of line 337. I didn't make the change because I was trying to minimize the changes I was making to the codebase to get a fix in (with mixed success).

Should I add a new declaration?